### ISSUE TITLE
Restructure APIs to remove versioning

### ISSUE THREAD

**Issue by rtyler** [lookout/deploydb#49]
@tommysullivan has convinced me to make a few changes to the API before we start having any API consumers:

 * `Accept` header for versioning API requests, e.g. `Accept: application/vnd.deploydb.v1+json`. Absence of this header will indicate that the "API default", which for now is going to be V1
 * Removing URL versions
 * All requests should respond with an `ETag` header (I don't really care about Etags, they're a PITA but it will allegedly make @tommysullivan happy)
 * Requests should respond with a correct `Last-Modified` header, in the case of database-backed resources this would be the `createdAt` value
 * Resource creations should use POST (just deferring to GitHub's convention, IDGAF)

The GitHub v3 API's concept of a "summary" and a "detailed" resource I'm on the fence about implementing since we've got so little deep data compared to GitHub.

I might add additional things in the future to this ticket

**Comment by tommysullivan** [comment]
ETag

Hey, we can use ETag to cache things but I am open to alternatives and also we can defer to a later version. It is mostly useful if it can save not only the network time between us and you but if it can also save you the server processing. If you have to go to SQL anyway and do a bunch of object hydration and what not, then the ETag will probably not save us much time.

I'd say its a nice to have.

**Comment by rtyler** [comment]
I threw the ETag stuff in because I'm *fairly* certain that we can just flip a setting in Dropwizard to get that for free, so why not!

**Comment by tommysullivan** [comment]
Accept Version

I think this will be nice. I am a fan of "YAGNI" so we can always start with something super generic like you identified - Accept: application/vnd.deploydb.v1+json and then make that the default. Then as we evolve certain representation formats we can expand on that. We don't need to build anything until we need it.

**Comment by tommysullivan** [comment]
Yeah usually you can cuz it will do a hash over the response body and that is good enough. But what Dropwizard won't do is if we do a HEAD or GET with a previously returned hash, it won't save you from calling through you DAOs down to your SQL stack. 

Side note: if you had a nosql json store behind the scenes it may come with built in hash management and know when its value is stale due to changes in dependency resources, etc...

**Comment by rtyler** [comment]
> Yeah usually you can cuz it will do a hash over the response body and that is good enough. But what Dropwizard won't do is if we do a HEAD or GET with a previously returned hash, it won't save you from calling through you DAOs down to your SQL stack

The hits to the JDBC (aka DAO) layer I'm not terribly concerned with, I like the ETag support more for preventing clients from needing to re-execute anything locally if I give them a "Not Modified" response

**Comment by tommysullivan** [comment]
re: POST / PUT and why github does what they do:

When you want to create a new resource but do not know a priori on the client what the URI of that resource should be, you POST to a uri that represents a "collection" or "container" type resource, and it generates a new URI, and returns that URI in the location header, so that you may in the future use GET with that URI to get updated representation. This does not preclude the POST from returning the representation itself, which can save the client a single call.

When you know a prior what the URI of the new resource is, or when you want to "replace" the resource completely, you PUT to that URI.

in action:

POST /people { "name": "Tommy" }
CREATED location: /people/12325122312
[optional representation of "Tommy" resource here]

GET /people/12325122312
[updated representation of Tommy]

GET /people
[gives a list with names and hrefs of all people in the collection]

---
vs.
---

PUT /people/tommy_sullivan { "name": "Tommy" } (if a previous representation existed, it is replaced with this one or an error is returned)
GET /people/tommy_sullivan 

the main difference is you PUT to the url of the newly created resource, you POST to the url of a separate resource, the collection, and get the url of the newly created resource in the location header.

apologies in advance if u knew this already :)

**Comment by tommysullivan** [comment]
yes, we will also like not having to process anything when ETAG helps us avoid doing so

**Comment by tommysullivan** [comment]
*not that it is any of my business how u implement under the covers* but curious, are u guys going to use stored procedures and then call those via jdbc using parameters and minimal prepared statements? or is it just jdbc and larger prepared statements without stored procs? or is it ORM? or something else? If ORM have you considered jooq? If jdbc + sprocs + minimal prepared statements this might be a nice way to insulate java from schema changes. if jdbc + larger prepared statements i have nothing to say. but i do hope it is not just straight sql strings and concatenation - hopefully you get some level of injection checking with whatever jdbc apis you are using...

**Comment by rtyler** [comment]
We're relying on Hibernate via [dropwizard's hibernate](http://dropwizard.io/manual/hibernate.html) which I'm fairly happy with.

**Comment by tommysullivan** [comment]
cool beans

**Comment by maheshkelkar** [comment]
Bumping up to 0.2.0, as discussed with Tyler "we can bump the issue to 0.2.0 but the thing I wanted to do in 0.1.0 is already there"
